File: packages/infrastructure/src/store/event-store.ts
import type Database from "better-sqlite3";
import type { PipelineEvent } from "@poria/core";

interface EventRow {
  seq: number;
  pipeline_id: string;
  kind: string;
  payload: string;
  created_at: string;
}

function rowToEvent(row: EventRow): PipelineEvent {
  const parsed = JSON.parse(row.payload) as Record<string, unknown>;
  // Restore Date objects from ISO strings
  if (typeof parsed["timestamp"] === "string") {
    parsed["timestamp"] = new Date(parsed["timestamp"] as string);
  }
  return parsed as unknown as PipelineEvent;
}

/**
 * Read-only event store. Writes happen via SqlitePipelineStore.saveStageTx().
 */
export class EventStore {
  private readonly selectByPipeline: Database.Statement;
  private readonly selectByKind: Database.Statement;

  constructor(db: Database.Database) {
    this.selectByPipeline = db.prepare(
      "SELECT * FROM events WHERE pipeline_id = ? ORDER BY seq ASC",
    );
    this.selectByKind = db.prepare(
      "SELECT * FROM events WHERE pipeline_id = ? AND kind = ? ORDER BY seq ASC",
    );
  }

  /**
   * Query all events for a pipeline, ordered by seq ascending.
   */
  queryByPipeline(pipelineId: string): PipelineEvent[] {
    const rows = this.selectByPipeline.all(pipelineId) as EventRow[];
    return rows.map(rowToEvent);
  }

  /**
   * Query events for a pipeline filtered by event kind.
   */
  queryByKind(pipelineId: string, kind: string): PipelineEvent[] {
    const rows = this.selectByKind.all(pipelineId, kind) as EventRow[];
    return rows.map(rowToEvent);
  }
}

File: crates/poria-infrastructure/src/store/event_store.rs
use chrono::Utc;
use rusqlite::{params, Connection};
use std::sync::Mutex;

use poria_core::pipeline::PipelineEvent;

pub struct EventStore {
    conn: Mutex<Connection>,
}

impl EventStore {
    pub fn new(conn: Connection) -> Self {
        Self {
            conn: Mutex::new(conn),
        }
    }

    pub fn query_by_pipeline(&self, pipeline_id: &str) -> Result<Vec<PipelineEvent>, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let mut stmt = conn
            .prepare("SELECT payload FROM events WHERE pipeline_id = ?1 ORDER BY seq ASC")
            .map_err(|e| e.to_string())?;

        let events: Vec<PipelineEvent> = stmt
            .query_map(params![pipeline_id], |row| {
                let payload: String = row.get(0)?;
                Ok(payload)
            })
            .map_err(|e| e.to_string())?
            .filter_map(|r| r.ok())
            .filter_map(|payload| serde_json::from_str(&payload).ok())
            .collect();

        Ok(events)
    }

    pub fn query_by_kind(
        &self,
        pipeline_id: &str,
        kind: &str,
    ) -> Result<Vec<PipelineEvent>, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let mut stmt = conn
            .prepare(
                "SELECT payload FROM events WHERE pipeline_id = ?1 AND kind = ?2 ORDER BY seq ASC",
            )
            .map_err(|e| e.to_string())?;

        let events: Vec<PipelineEvent> = stmt
            .query_map(params![pipeline_id, kind], |row| {
                let payload: String = row.get(0)?;
                Ok(payload)
            })
            .map_err(|e| e.to_string())?
            .filter_map(|r| r.ok())
            .filter_map(|payload| serde_json::from_str(&payload).ok())
            .collect();

        Ok(events)
    }

    pub fn append(&self, pipeline_id: &str, event: &PipelineEvent) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let payload = serde_json::to_string(event).map_err(|e| e.to_string())?;
        let kind = serde_json::from_str::<serde_json::Value>(&payload)
            .ok()
            .and_then(|v| v.get("kind").and_then(|k| k.as_str()).map(String::from))
            .unwrap_or_else(|| "unknown".into());

        conn.execute(
            "INSERT INTO events (pipeline_id, kind, payload, created_at) VALUES (?1, ?2, ?3, ?4)",
            params![pipeline_id, kind, payload, Utc::now().to_rfc3339()],
        )
        .map_err(|e| e.to_string())?;

        Ok(())
    }
}
